### PR TITLE
Make API compatible with SQL backend

### PR DESCRIPTION
**Pull request by NoahCarnahan**
@dimagi/scale-team cc @snopoke 

Best reviewed commit-by-commit

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/api/resources/v0_1.py, line 10 - F401 'XFormNotFound' imported but unused
* corehq/apps/api/resources/v0_1.py, line 11 - F401 'FormAccessors' imported but unused
* corehq/apps/api/resources/v0_1.py, line 43 - F401 'object_does_not_exist' imported but unused
* corehq/apps/api/resources/v0_4.py, line 217 - E302 expected 2 blank lines, found 1
* corehq/form_processor/models.py, line 576 - E303 too many blank lines (2)

File: corehq/apps/api/resources/v0_4.py [hunk 1]
```diff
@@ -216,10 +216,10 @@ class CommCareCaseResource(SimpleSortableResourceMixin, v0_3.CommCareCaseResourc
     ))
 
     child_cases = UseIfRequested(ToManyDictField('corehq.apps.api.resources.v0_4.CommCareCaseResource',
-                                                 attribute=lambda case: dict([ (index.identifier, CommCareCase.get(index.referenced_id)) for index in case.indices])))
+                                                 attribute=lambda case: dict([ (index.identifier, CommCareCase.get(index.referenced_id)) for index in case.reverse_indicies])))
 
     parent_cases = UseIfRequested(ToManyDictField('corehq.apps.api.resources.v0_4.CommCareCaseResource',
-                                                  attribute=lambda case: dict([ (index.identifier, CommCareCase.get(index.referenced_id)) for index in case.reverse_indices])))
+                                                  attribute=lambda case: dict([ (index.identifier, CommCareCase.get(index.referenced_id)) for index in case.indices])))
```
Comment (NoahCarnahan) [inline corehq/apps/api/resources/v0_4.py]: Apparently the parent_cases field was returning the child cases (and vice versa) for some time. These fields don't seem to be documented on the confluence page though (and you need to add `?parent_cases__full=true` to the url to get it), so hopefully no one has been relying on the erroneous behavior.

Comment (benrudolph) [inline corehq/apps/api/resources/v0_4.py]: haha :flushed:

File: corehq/apps/api/resources/v0_3.py [hunk 2]
```diff
@@ -68,22 +69,11 @@ class XFormInstanceResource(v0_1.XFormInstanceResource, DomainSpecificResourceMi
...
-        # Logic borrowed from util.get_object_or_not_exist
+        instance_id = kwargs['pk']
         try:
-            doc = couchforms.fetch_and_wrap_form(doc_id)
-            if doc and doc.domain == domain:
-                return doc
-        except (ResourceNotFound, couchforms.UnexpectedDeletedXForm):
-            pass  # covered by the below
-        except AttributeError:
-            # there's a weird edge case if you reference a form with a case id
```
Comment (benrudolph) [inline corehq/apps/api/resources/v0_3.py]: lol

File: corehq/ex-submodules/casexml/apps/case/xform.py [hunk 3]
```diff
@@ -366,16 +366,14 @@ def get_case_ids_from_form(xform):
...
-    cases = [CommCareCase.wrap(doc)
-             for doc in iter_docs(CommCareCase.get_db(), case_ids)]
-
     domain = get_and_check_xform_domain(xform)
-    if domain:
-        for case in cases:
-            assert case.domain == domain
-
-    return cases
+    case_accessor = CaseAccessors(domain)
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/xform.py]: did you find instances when we don't have the domain? seems bad

File: corehq/form_processor/abstract_models.py [hunk 4]
```diff
@@ -155,6 +155,65 @@ def get_subcases(self):
     def get_case_property(self, property):
         raise NotImplementedError
 
+    def get_json(self, lite=False):
```
Comment (benrudolph) [inline corehq/form_processor/abstract_models.py]: we should use our serializers for this if possible: https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/serializers.py

**Comment by benrudolph** [comment]
:candy:

File: corehq/apps/api/resources/v0_4.py (hunk 1 above)
Comment (snopoke) [inline corehq/apps/api/resources/v0_4.py]: Should we add docs?

Comment (NoahCarnahan) [inline corehq/apps/api/resources/v0_4.py]: Yeah, I'll update the confluence page once this is deployed.

File: corehq/ex-submodules/casexml/apps/case/xform.py (hunk 3 above)
Comment (NoahCarnahan) [inline corehq/ex-submodules/casexml/apps/case/xform.py]: No I didn't find instances of that happening. This function seems like it had been written to account for instances when domain is `None`, so thought I would preserve that behavior.

File: corehq/form_processor/abstract_models.py (hunk 4 above)
Comment (NoahCarnahan) [inline corehq/form_processor/abstract_models.py]: Good point.
After reviewing this comment I realized it was actually overkill to move this whole function to `AbstractCommCareCase`, so I'm going to move it back to `CommCareCase`, and just use tastypie's serialization workflow to grab the bits it needs (properties and indices) while leaving `get_json()` unchanged: https://github.com/dimagi/commcare-hq/pull/10976/commits/e84e725d6241b167966d2262c3e2d96004a6fbde

Comment (benrudolph) [inline corehq/form_processor/abstract_models.py]: :+1:

**Comment by benrudolph** [comment]
test failure real:
```
======================================================================
FAIL: testCaseAPIResultJSON (corehq.apps.cloudcare.tests.test_api.CaseAPITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/cloudcare/tests/test_api.py", line 233, in testCaseAPIResultJSON
    self.assertEqual(json['properties']['case_type'], None)
AssertionError: '' != None
```

**Comment by NoahCarnahan** [comment]
:white_check_mark:

**Pull request merged**